src/Tasks/AssignTargetPath.cs

```diff
@@ -71,13 +71,20 @@ public override bool Execute()
 
                 for (int i = 0; i < Files.Length; ++i)
                 {
-                    string link = Files[i].GetMetadata(ItemMetadataNames.link);
                     AssignedFiles[i] = new TaskItem(Files[i]);
+                    string targetPath = Files[i].GetMetadata(ItemMetadataNames.targetPath);
 
-                    // If file has a link, use that.
-                    string targetPath = link;
+                    // If TargetPath is already set, copy it over.
+                    // https://github.com/dotnet/msbuild/issues/2795
+                    if (!string.IsNullOrEmpty(targetPath))
+                    {
+                        AssignedFiles[i].SetMetadata(ItemMetadataNames.targetPath, EscapingUtilities.Escape(targetPath));
```

Comment:
nit: Instead of the continue and duplication you could let it fall through. First try TargetPath. If null or empty try Link. If null or empty run that if block. Then set the escaped value and return.